util/dvsim/utils.py

```diff
@@ -86,84 +95,280 @@ def parse_hjson(hjson_file):
-            # of ignored_wildcards list or if substitutions for them were not
-            # found, then, we return. If all substitutions were made, then check
-            # for second level of indirection:
-            #
-            # For example: lets say we supply the dict:
-            # {var: '{{foo}_xyz{bar}}', foo: p, bar: q, p_xyz_q: baz}
-            #
-            # Then after the substitutions above: {var: '{p_xyz_q}', ...}
-            # We need to now substitute {p_xyz_q}, so that the final value of
-            # var is 'baz'.
-            if not (ignored_wildcards_found or no_substitutions_found):
-                var = subst_wildcards(var, mdict, ignored_wildcards,
-                                      ignore_error)
-    return var
+    wildcard_re = re.compile(r"{([A-Za-z0-9\_]+)}")
+
+    # Work from left to right, expanding each wildcard we find. idx is where we
+    # should start searching (so that we don't keep finding a wildcard that
+    # we've decided to ignore).
+    idx = 0
+
+    any_err = False
+
+    while True:
+        right_str = var[idx:]
+        match = wildcard_re.search(right_str)
+
+        # If no match, we're done.
+        if match is None:
+            return (var, any_err)
+
+        name = match.group(1)
+
+        # If the name should be ignored, skip over it.
+        if name in ignored:
+            idx += match.end()
+            continue
+
+        # If the name has been seen already, we've spotted circular recursion.
+        # If we're ignoring errors, skip over it. Otherwise, error.
+        if name in seen:
+            if ignore_error:
+                idx += match.end()
+                any_err = True
+                continue
+
+            raise ValueError('String contains circular expansion of '
+                             'wildcard {!r}.'
+                             .format(match.group(0)))
+
+        # Treat eval_cmd specially
+        if name == 'eval_cmd':
+            cmd = _subst_wildcards(right_str[match.end():],
+                                   mdict, ignored, ignore_error, seen)[0]
+
+            # Are there any wildcards left in cmd? If not, we can run the
+            # command and we're done.
+            cmd_matches = list(wildcard_re.finditer(cmd))
+            if not cmd_matches:
+                var = var[:idx] + run_cmd(cmd)
+                continue
+
+            # Otherwise, check that each of them is ignored, or that
+            # ignore_error is True.
+            bad_names = False
+            if not ignore_error:
+                for cmd_match in cmd_matches:
+                    if cmd_match.group(1) not in ignored:
+                        bad_names = True
+
+            if bad_names:
+                raise ValueError('Cannot run eval_cmd because the command '
+                                 'expands to {!r}, which still contains a '
+                                 'wildcard.'
+                                 .format(cmd))
```

Comment:
Is this to avoid defining the `bad_names` variable? I'm not convinced it's clearer (and the error message gets indented even further right, which means even more wrapping).

Reply:
Nvm, existing approach is fine.